uuencode/uuencode.1
.Dd December 2008
.Dt UUENCODE 1

.Sh NAME

.Nm uuencode
.Nd encode a binary file

.Sh SYNOPSIS

.Nm uuencode
.Ar [-m][file] decode_pathname

.Sh DESCRIPTION

    The uuencode utility shall write an encoded version of the named input
file, or standard input if no file is specified, to standard output. The
output shall be encoded using one of the algorithms described in the STDOUT
section and shall include the file access permission bits (in chmod octal or
symbolic notation) of the input file and the decode_pathname, for re-creation
of the file on another system that conforms to this volume of IEEE Std
1003.1-2001.

.Sh OPTIONS

    The uuencode utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following option shall be supported by the implementation:

    -m
        Encode the output using the MIME Base64 algorithm described in
STDOUT. If -m is not specified, the historical algorithm described in STDOUT
shall be used.

.Sh OPERANDS

    The following operands shall be supported:

    decode_pathname
        The pathname of the file into which the uudecode utility shall place
the decoded file. Specifying a decode_pathname operand of /dev/stdout shall
indicate that uudecode is to use standard output. If there are characters in
decode_pathname that are not in the portable filename character set the
results are unspecified.
    file
        A pathname of the file to be encoded.

.Sh STDIN

    See the INPUT FILES section.

.Sh INPUT FILES

    Input files can be files of any type.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of
uuencode:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    uuencode Base64 Algorithm

    The standard output shall be a text file (encoded in the character set of
the current locale) that begins with the line:

    "begin-base64%s%s\n", <mode>, <decode_pathname>

    and ends with the line:

    "====\n"

    In both cases, the lines shall have no preceding or trailing <blank>s.

    The encoding process represents 24-bit groups of input bits as output
strings of four encoded characters. Proceeding from left to right, a 24-bit
input group shall be formed by concatenating three 8-bit input groups. Each
24-bit input group then shall be treated as four concatenated 6-bit groups,
each of which shall be translated into a single digit in the Base64 alphabet.
When encoding a bit stream via the Base64 encoding, the bit stream shall be
presumed to be ordered with the most-significant bit first. That is, the
first bit in the stream shall be the high-order bit in the first byte, and
the eighth bit shall be the low-order bit in the first byte, and so on. Each
6-bit group is used as an index into an array of 64 printable characters, as
shown in uuencode Base64 Values.
    Table: uuencode Base64 Values

    Value
    	

    Encoding
    	

    Value
    	

    Encoding
    	

    Value
    	

    Encoding
    	

    Value
    	

    Encoding
    	

     
    	

     
    	

     

    0
    	

    A
    	

    17
    	

    R
    	

    34
    	

    i
    	

    51
    	

    z
    	

     
    	

     
    	

     

    1
    	

    B
    	

    18
    	

    S
    	

    35
    	

    j
    	

    52
    	

    0
    	

     
    	

     
    	

     

    2
    	

    C
    	

    19
    	

    T
    	

    36
    	

    k
    	

    53
    	

    1
    	

     
    	

     
    	

     

    3
    	

    D
    	

    20
    	

    U
    	

    37
    	

    l
    	

    54
    	

    2
    	

     
    	

     
    	

     

    4
    	

    E
    	

    21
    	

    V
    	

    38
    	

    m
    	

    55
    	

    3
    	

     
    	

     
    	

     

    5
    	

    F
    	

    22
    	

    W
    	

    39
    	

    n
    	

    56
    	

    4
    	

     
    	

     
    	

     

    6
    	

    G
    	

    23
    	

    X
    	

    40
    	

    o
    	

    57
    	

    5
    	

     
    	

     
    	

     

    7
    	

    H
    	

    24
    	

    Y
    	

    41
    	

    p
    	

    58
    	

    6
    	

     
    	

     
    	

     

    8
    	

    I
    	

    25
    	

    Z
    	

    42
    	

    q
    	

    59
    	

    7
    	

     
    	

     
    	

     

    9
    	

    J
    	

    26
    	

    a
    	

    43
    	

    r
    	

    60
    	

    8
    	

     
    	

     
    	

     

    10
    	

    K
    	

    27
    	

    b
    	

    44
    	

    s
    	

    61
    	

    9
    	

     
    	

     
    	

     

    11
    	

    L
    	

    28
    	

    c
    	

    45
    	

    t
    	

    62
    	

    +
    	

     
    	

     
    	

     

    12
    	

    M
    	

    29
    	

    d
    	

    46
    	

    u
    	

    63
    	

    /
    	

     
    	

     
    	

     

    13
    	

    N
    	

    30
    	

    e
    	

    47
    	

    v
    	

     
    	

     
    	

     
    	

     
    	

     

    14
    	

    O
    	

    31
    	

    f
    	

    48
    	

    w
    	

    (pad)
    	

    =
    	

     
    	

     
    	

     

    15
    	

    P
    	

    32
    	

    g
    	

    49
    	

    x
    	

     
    	

     
    	

     
    	

     
    	

     

    16
    	

    Q
    	

    33
    	

    h
    	

    50
    	

    y
    	

     
    	

     
    	

     
    	

     
    	

     

    The character referenced by the index shall be placed in the output
string.

    The output stream (encoded bytes) shall be represented in lines of no
more than 76 characters each. All line breaks or other characters not found
in the table shall be ignored by decoding software (see uudecode).

    Special processing shall be performed if fewer than 24 bits are available
at the end of a message or encapsulated part of a message. A full encoding
quantum shall always be completed at the end of a message. When fewer than 24
input bits are available in an input group, zero bits shall be added (on the
right) to form an integral number of 6-bit groups. Output character positions
that are not required to represent actual input data shall be set to the
character '='. Since all Base64 input is an integral number of octets, only
the following cases can arise:

        The final quantum of encoding input is an integral multiple of 24
bits; here, the final unit of encoded output shall be an integral multiple of
4 characters with no '=' padding.

        The final quantum of encoding input is exactly 16 bits; here, the
final unit of encoded output shall be three characters followed by one '='
padding character.

        The final quantum of encoding input is exactly 8 bits; here, the
final unit of encoded output shall be two characters followed by two '='
padding characters.

    A terminating "====" evaluates to nothing and denotes the end of the
encoded data.
    uuencode Historical Algorithm

    The standard output shall be a text file (encoded in the character set of
the current locale) that begins with the line:

    "begin%s%s\n" <mode>, <decode_pathname>

    and ends with the line:

    "end\n"

    In both cases, the lines shall have no preceding or trailing <blank>s.

    The algorithm that shall be used for lines in between begin and end takes
three octets as input and writes four characters of output by splitting the
input at six-bit intervals into four octets, containing data in the lower six
bits only. These octets shall be converted to characters by adding a value of
0x20 to each octet, so that each octet is in the range [0x20,0x5f], and then
it shall be assumed to represent a printable character in the ISO/IEC
646:1991 standard encoded character set. It then shall be translated into the
corresponding character codes for the codeset in use in the current locale.
(For example, the octet 0x41, representing 'A', would be translated to 'A' in
the current codeset, such as 0xc1 if it were EBCDIC.)

    Where the bits of two octets are combined, the least significant bits of
the first octet shall be shifted left and combined with the most significant
bits of the second octet shifted right. Thus the three octets A, B, C shall
be converted into the four octets:

    0x20 + (( A >> 2                    ) & 0x3F)
    0x20 + (((A << 4) | ((B >> 4) & 0xF)) & 0x3F)
    0x20 + (((B << 2) | ((C >> 6) & 0x3)) & 0x3F)
    0x20 + (( C                         ) & 0x3F)

    These octets then shall be translated into the local character set.

    Each encoded line contains a length character, equal to the number of
characters to be decoded plus 0x20 translated to the local character set as
described above, followed by the encoded characters. The maximum number of
octets to be encoded on each line shall be 45.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    The file is expanded by 35 percent (each three octets become four, plus
control information) causing it to take longer to transmit.

    Since this utility is intended to create files to be used for data
interchange between systems with possibly different codesets, and to
represent binary data as a text file, the ISO/IEC 646:1991 standard was
chosen for a midpoint in the algorithm as a known reference point. The output
from uuencode is a text file on the local system. If the output were in the
ISO/IEC 646:1991 standard codeset, it might not be a text file (at least
because the <newline>s might not match), and the goal of creating a text file
would be defeated. If this text file was then carried to another machine with
the same codeset, it would be perfectly compatible with that system's
uudecode. If it was transmitted over a mail system or sent to a machine with
a different codeset, it is assumed that, as for every other text file, some
translation mechanism would convert it (by the time it reached a user on the
other system) into an appropriate codeset. This translation only makes sense
from the local codeset, not if the file has been put into a ISO/IEC 646:1991
standard representation first. Similarly, files processed by uuencode can be
placed in pax archives, intermixed with other text files in the same codeset.

.Sh EXAMPLES

    None.

.Sh RATIONALE

    A new algorithm was added at the request of the international community
to parallel work in RFC 2045 (MIME). As with the historical uuencode format,
the Base64 Content-Transfer-Encoding is designed to represent arbitrary
sequences of octets in a form that is not humanly readable. A 65-character
subset of the ISO/IEC 646:1991 standard is used, enabling 6 bits to be
represented per printable character. (The extra 65th character, '=', is used
to signify a special processing function.)

    This subset has the important property that it is represented identically
in all versions of the ISO/IEC 646:1991 standard, including US ASCII, and all
characters in the subset are also represented identically in all versions of
EBCDIC. The historical uuencode algorithm does not share this property, which
is the reason that a second algorithm was added to the ISO POSIX-2 standard.

    The string "====" was used for the termination instead of the end used in
the original format because the latter is a string that could be valid
encoded input.

    In an early draft, the -m option was named -b (for Base64), but it was
renamed to reflect its relationship to the RFC 2045. A -u was also present to
invoke the default algorithm, but since this was not historical practice, it
was omitted as being unnecessary.

    See the RATIONALE section in uudecode for the derivation of the
/dev/stdout symbol.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    chmod(), mailx, uudecode



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



uuencode/uuencode.cc
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/uio.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <libpu.h>

#define PFX "uuencode: "

static const char doc[] =
N_("uuencode - encode a binary file");

static const char args_doc[] = N_("[file] decode_pathname");

static struct argp_option options[] = {
	{ "base64", 'm', NULL, 0,
	  N_("Encode the output using the MIME Base64 algorithm") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

enum random_constants {
	UUE_BUF_SZ		= 4096,
	CHUNK_SZ		= 3,
	S_IXALL			= S_IRWXU | S_IRWXG | S_IRWXO,

	LN_MAX			= 45
};

static unsigned char buf[UUE_BUF_SZ];
static unsigned char outbuf[UUE_BUF_SZ];
static int outbuf_len;
static unsigned char *line_start;

static int line_len;
static int line_bytes;

static unsigned char spill[4];
static int n_spill;

static int opt_base64;
static const unsigned char *tbl;

static const unsigned char tbl_base64[64] = {
	'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
	'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
	'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd',
	'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n',
	'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x',
	'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7',
	'8', '9', '+', '/'
};

static const unsigned char tbl_trad[64] = {
	'`',
	'!', '"', '#', '$', '%', '&', '\'', '(', ')', '*',
	'+', ',', '-', '.', '/', '0', '1', '2', '3', '4',
	'5', '6', '7', '8', '9', ':', ';', '<', '=', '>',
	'?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
	'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
	'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\',
	']', '^', '_'
};

static void flush_outbuf(void)
{
	if (outbuf_len > 0) {
		write_buf(outbuf, outbuf_len);
		outbuf_len = 0;
	}
}

static void *push_outbuf(const void *s, size_t len)
{
	void *mem;
	if ((outbuf_len + len) > UUE_BUF_SZ)
		flush_outbuf();

	mem = &outbuf[outbuf_len];

	memcpy(mem, s, len);
	outbuf_len += len;

	return mem;
}

static void push_line (void)
{
	if (!opt_base64)
		*line_start = tbl_trad[line_bytes];

	push_outbuf("\n", 1);

	line_len = 0;
	line_bytes = 0;
	if (!opt_base64)
		line_start = (unsigned char *) push_outbuf(" ", 1);
}

static void encode_chunk(const unsigned char *s)
{
	unsigned char c[4];
	int ci[4], i;

	/* divide 3 bytes into four 6-bit chunks, reading
	 * from left (MSB) to right
	 */
	ci[0] = ((s[0] >> 2) & 0x3f);

	ci[1] = ((s[0] << 4) & 0x3f);
	ci[1] |= ((s[1] >> 4) & 0xf);

	ci[2] = ((s[1] << 2) & 0x3f);
	ci[2] |= ((s[2] >> 6) & 0x3);

	ci[3] = (s[2] & 0x3f);

	for (i = 0; i < 4; i++)
		c[i] = tbl[ci[i]];

	push_outbuf(c, sizeof(c));
	line_len += sizeof(c);
	line_bytes += CHUNK_SZ;

	if (line_bytes >= LN_MAX)
		push_line();
}

static void finalize_base64(void)
{
	unsigned char c[4];

	c[0] = tbl_base64[ spill[0] >> 2 ];

	if (n_spill == 1) {
		c[1] = tbl_base64[ ((spill[0] & 0x3) << 4) ];
		c[2] = '=';
	} else {
		c[1] = tbl_base64[ ((spill[0] & 0x3) << 4) | (spill[1] >> 4) ];
		c[2] = tbl_base64[ ((spill[1] & 0xf) << 2) ];
	}

	c[3] = '=';

	push_outbuf(c, sizeof(c));
	line_len += sizeof(c);
	line_bytes += n_spill;
}

static void finalize_traditional(void)
{
	memset(&spill[n_spill], 0, CHUNK_SZ - n_spill);
	encode_chunk(spill);
	n_spill = 0;
}

static void flush_last_line (void)
{
	if (n_spill) {
		if (opt_base64)
			finalize_base64();
		else
			finalize_traditional();
	}

	push_line();

	if (!opt_base64) {
		*line_start = '`';
		push_outbuf("\n", 1);
	}
}

static void push_bytes(ssize_t len)
{
	const unsigned char *s = buf;

	if (n_spill) {
		while ((n_spill < CHUNK_SZ) && (len > 0)) {
			spill[n_spill] = *s;
			s++;
			len--;
			n_spill++;
		}

		if (n_spill < CHUNK_SZ)
			return;

		encode_chunk(spill);
		n_spill = 0;
	}

	while (len >= CHUNK_SZ) {
		encode_chunk(s);
		s += CHUNK_SZ;
		len -= CHUNK_SZ;
	}

	if (len > 0) {
		memcpy(spill, s, len);
		n_spill = len;
	}
}

static void do_header(int fd, const char *fn_string)
{
	struct stat st;
	char linebuf[PATH_MAX + 20];

	if (fstat(fd, &st) == 0)
		st.st_mode &= S_IXALL;
	else
		st.st_mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;

	snprintf(linebuf, sizeof(linebuf), "begin%s %03o %s\n",
		 opt_base64 ? "-base64" : "", st.st_mode, fn_string);
	push_outbuf(linebuf, strlen(linebuf));
}

static int do_encode(int fd, const char *fn_string)
{
	ssize_t rc;

	if (opt_base64)
		tbl = tbl_base64;
	else
		tbl = tbl_trad;

	do_header(fd, fn_string);

	if (!opt_base64)
		line_start = (unsigned char *) push_outbuf(" ", 1);
	line_len = 0;
	line_bytes = 0;

	while (1) {
		rc = read(fd, buf, sizeof(buf));
		if (rc == 0)
			break;
		if (rc < 0) {
			perror(fn_string);
			return 1;
		}

		push_bytes(rc);
	}

	flush_last_line();

	if (opt_base64)
		push_outbuf("====\n", 5);
	else
		push_outbuf("end\n", 6);

	flush_outbuf();

	return 0;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case 'm':
		opt_base64 = 1;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

int main (int argc, char *argv[])
{
	char *input_fn = NULL, *fn_name;
	int fd, rc;
	int idx = -1;

	pu_init();

	error_t argp_rc = argp_parse(&argp, argc, argv, 0, &idx, NULL);
	if (argp_rc || (idx < 0)) {
		fprintf(stderr, _("%s: argp_parse failed: %s\n"),
			argv[0], strerror(argp_rc));
		return EXIT_FAILURE;
	}

	if ((argc - idx) == 1)
		fn_name = argv[idx];
	else if ((argc - idx) == 2) {
		input_fn = argv[idx++];
		fn_name = argv[idx];
	}
	else {
		fprintf(stderr, PFX "invalid arguments\n");
		return 1;
	}

	if (input_fn) {
		fd = open(input_fn, O_RDONLY);
		if (fd < 0) {
			perror(input_fn);
			return 1;
		}

		/* Do we care about the return value of posix_fadvise(2) ? */
		posix_fadvise(fd, 0, 0, POSIX_FADV_SEQUENTIAL);
	} else {
		fd = STDIN_FILENO;
	}

	rc = do_encode(fd, fn_name);

	if (input_fn) {
		if (close(fd) < 0)
			rc |= 1;
	}

	return rc;
}

